File: LeetCode/129. Sum Root to Leaf Numbers.py
# https://leetcode.com/problems/sum-root-to-leaf-numbers/description/


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sumNumbers(self, node: TreeNode | None, current_sum: str = "") -> int:
        if node is None:
            return 0

        if node.left is None and node.right is None:
            return int(current_sum + str(node.val))

        return self.sumNumbers(
            node.left, current_sum + str(node.val)
        ) + self.sumNumbers(node.right, current_sum + str(node.val))

File: LeetCode/129. Sum Root to Leaf Numbers.js
// https://leetcode.com/problems/sum-root-to-leaf-numbers/description/

// const sumNumbers = (root) => {
//   const inner = (node, sum) => {
//     if (!node) {
//       return 0;
//     }

//     if (!node.left && !node.right) {
//       return Number(sum + node.val);
//     }

//     return inner(node.left, sum + node.val) + inner(node.right, sum + node.val);
//   };

//   return inner(root, '');
// };

const sumNumbers = (node, sum = '') => {
  if (!node) {
    return 0;
  }

  if (!node.left && !node.right) {
    return Number(sum + node.val);
  }

  return sumNumbers(node.left, sum + node.val) + sumNumbers(node.right, sum + node.val);
};
